test/testutils/util.go

```diff
@@ -696,3 +738,194 @@ func SetLwsPartition(ctx context.Context, k8sClient client.Client, lws *leaderwo
+		gomega.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: leaderPodName, Namespace: lws.Namespace}, leaderPod)).To(gomega.Succeed())
+
+		// Get the PodGroup using unstructured object
+		podGroup := &unstructured.Unstructured{}
+		podGroup.SetGroupVersionKind(gvk)
+		gomega.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: expectedPgName, Namespace: lws.Namespace}, podGroup)).To(gomega.Succeed())
+
+		// Validate podgroup owner reference and spec
+		validateOwnerReference(podGroup, leaderPod)
+		validatePodGroupSpec(podGroup, lws)
+	}
+}
+
+// UpdatePodGroupAtIndex manually updates the PodGroup for a specific group index
+func UpdatePodGroupAtIndex(ctx context.Context, k8sClient client.Client, providerType schedulerprovider.ProviderType, schedulerProvider schedulerprovider.SchedulerProvider, lws *leaderworkerset.LeaderWorkerSet, groupIndex string) {
+	gvk := getPodGroupGVK(providerType)
+	if gvk.Empty() {
+		ginkgo.Skip("Unsupported scheduler provider for PodGroup update")
+		return
+	}
+
+	// Get revisions for PodGroup naming
+	var leaderSts appsv1.StatefulSet
+	gomega.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &leaderSts)).To(gomega.Succeed())
+	revision := revisionutils.GetRevisionKey(&leaderSts)
+
+	// Manually simulate replica update: delete old leader Pod and PodGroup, create new ones
+	// Delete old leader pod
+	oldPod := &corev1.Pod{}
+	podName := fmt.Sprintf("%s-%s", lws.Name, groupIndex)
+	if err := k8sClient.Get(ctx, types.NamespacedName{Name: podName, Namespace: lws.Namespace}, oldPod); err == nil {
+		gomega.Expect(k8sClient.Delete(ctx, oldPod)).To(gomega.Succeed())
+	}
+
+	// Delete old PodGroup (simulating garbage collection) using unstructured object
+	oldPgName := schedulerprovider.GetPodGroupName(lws.Name, groupIndex, revision)
+	oldPodGroup := &unstructured.Unstructured{}
+	oldPodGroup.SetGroupVersionKind(gvk)
+	if err := k8sClient.Get(ctx, types.NamespacedName{Name: oldPgName, Namespace: lws.Namespace}, oldPodGroup); err == nil {
+		gomega.Expect(k8sClient.Delete(ctx, oldPodGroup)).To(gomega.Succeed())
+	}
+
+	// Create new pod (this will trigger new PodGroup creation)
+	startIndex, err := strconv.Atoi(groupIndex)
+	gomega.Expect(err).ToNot(gomega.HaveOccurred())
+	injectFn := func(pod *corev1.Pod) {
+		gomega.Expect(schedulerProvider).NotTo(gomega.BeNil())
+		err = schedulerProvider.InjectPodGroupMetadata(pod)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}
+	gomega.Expect(CreateLeaderPodsWithInjectFn(ctx, leaderSts, k8sClient, lws, startIndex, startIndex+1, injectFn)).To(gomega.Succeed())
+}
+
+// ExpectValidPodGroupAtIndex checks that a PodGroup exists for the specified group index and has the correct owner reference
+func ExpectValidPodGroupAtIndex(ctx context.Context, k8sClient client.Client, provider schedulerprovider.ProviderType, lws *leaderworkerset.LeaderWorkerSet, groupIndex string) {
+	gvk := getPodGroupGVK(provider)
+	if gvk.Empty() {
```

Comment:
ditto

Reply:
OK I have changed to `ginkgo.Fail`, please check